cpp/ql/lib/semmle/code/cpp/models/implementations/Memcpy.qll

```diff
@@ -106,6 +106,8 @@ private class MemcpyFunction extends ArrayFunction, DataFlowFunction, SideEffect
     not this.hasGlobalName(["bcopy", mempcpy(), "memccpy"]) and
     index = this.getParamDest()
   }
+
+  override TCxxException getExceptionType() { any() }
```

Comment:
I find this confusing. Together with the `NonThrowing` this seems to say that:

> This function is non-throwing, but only in the case of C++ exceptions. So it may still throw a SEH exception.

Is that the correct reading?

Reply:
That's right. That's what it is saying. It is no longer sufficient to say a function doesn't throw, you have to say how it doesn't throw (which kind of exception doesn't it throw). If they want to say it doesn't throw any you can just return the parent exception type. 

The issue we got into with making memcpy nonthrowing is that it is true that it doesn't throw a C++ exception, but it absolutely throws a SEH exception.

Reply:
Thanks.

I wonder if this is all somewhat overly complicated. In my understanding the following cases are interesting:

* C++ exceptions:
  * Functions that may throw (this is the default, but are modelling of this in the IR is limited)
  * Functions that do **never** throw a C++ exception, i.e., functions marked either as `noexcept or that are C functions (that we model)
* SEH exceptions:
  * Functions that may throw (this is the default, and we want to improve the modelling in the IR here)
  * Functions that **always** throw

Is this correct?

Reply:
Overhauled the PR, let me know if that works.

Reply:
> Introduce a new class AlwaysSehThrowingFunction which is used to model functions that always throw an SEH exception.

When you say "always throw an SEH exception" does that mean that the function will _always_ throw an SEH exception when called or that the function can _only_ throw an SEH exception but doesn't necessarily always do so? To me the name sounds like the former but I guess it's actually the latter?

Reply:
The former: the function will _always_ throw an SEH exception

Reply:
Thanks. I see. So I guess it's for functions whose purpose is to raise exceptions? Like something along the lines of `throw_my_SEH_exception`?

Reply:
Yes: `RaiseException`, `ExRaiseAccessViolation`, `ExRaiseDatatypeMisalignment`.

Reply:
Alright, thanks :)